### ISSUE TITLE
Training instability for weather setting, also reproduced when trying the provided Docker environment

### ISSUE THREAD

**Issue by YuKDseele** [ZeldaM1/All-in-one#3]
Hi, thanks for releasing this project.

I am trying to reproduce the weather training setting, but I ran into training instability very early, and I also found the training recipe a bit unclear. I wanted to ask whether some required pretraining / staged training step is missing from the public instructions.

### What I tried
I used the weather S config with the training script:

./tools/dist_train_iter.sh configs/weather/all_in_one_S.py 2 "--train_dataset data/all-in-one-train --test_dataset data/all-in-one-test --exp_name all_in_one_S_weather"

I also tried following the Docker-based route mentioned in the repo. However, even when switching to the provided Docker environment path, I still could not get a clean end-to-end reproduction workflow, so I am not sure whether I am missing an undocumented prerequisite.

### Main issue during training
When training starts, it quickly becomes numerically unstable and repeatedly prints:

non-finite norm, skip this batch

The early printed statistics look like:

{'bpp_y': '7.53791 7.53791 ', 'bpp_z': '0.12869 0.12869 ', 'mse': '1.14728 1.14728 ', 'ssim': '0.00000 0.00000 ', 'costs': ('154.51787 377.70859 ',), 'losses': ('154.51787 377.70859 ',)}
{'bpp_y': '5.36227 5.36227 ', 'bpp_z': '0.12861 0.12861 ', 'mse': '0.58286 0.58286 ', 'ssim': '0.00000 0.00000 ', 'costs': ('80.09702 193.48657 ',), 'losses': ('80.09702 193.48657 ',)}

{'bpp_y': '5.47628 5.47628 ', 'bpp_z': '0.12859 0.12859 ', 'mse': '0.41654 0.41654 ', 'ssim': '0.00000 0.00000 ', 'costs': ('344.15039 858.68610 ',), 'losses': ('344.15039 858.68610 ',)}
{'bpp_y': '7.65921 7.65921 ', 'bpp_z': '0.12867 0.12867 ', 'mse': '0.74923 0.74923 ', 'ssim': '0.00000 0.00000 ', 'costs': ('616.72650 1542.21704 ',), 'losses': ('616.72650 1542.21704 ',)}
non-finite norm, skip this batch
non-finite norm, skip this batch
non-finite norm, skip this batch
...

### Why I am confused
From the released configs, I noticed that the weather config contains:

resume_from = f'./workdirs/{exp_name}/iter_150000.pth'

This makes me wonder whether the public all_in_one_S.py / all_in_one_L.py configs are not actually intended to be trained directly from scratch, but instead assume an earlier stage/checkpoint.

Also, the repo contains multiple base weather configs with different patch sizes (256, 320, 384), so I am not sure whether the intended recipe is a progressive training schedule such as:

- first train with 256 patch size
- then continue with 320
- then continue with 384

### Questions
1. Are configs/weather/all_in_one_S.py and configs/weather/all_in_one_L.py intended to be run directly from scratch?
2. Or do they require resuming from an earlier stage that is not fully described in the repo?
3. Is the intended weather training procedure progressive (256 -> 320 -> 384)?
4. Are the repeated identical pairs in the printed stats expected at the beginning due to multi-rate training?
5. Is there an official recommended reproduction path using the provided Docker environment, including the exact training order / resume checkpoints?

If possible, it would be very helpful to document the full training pipeline more explicitly, especially for users trying to reproduce the weather setting from scratch.

Thanks.

**Comment by ZeldaM1** [comment]
Hi,

Thanks for your interest in our work!  

1. all_in_one_S.py/all_in_one_L.py files are configurations of the final stage.  We adopt a progressive training strategy (Sec. 3.4), so later stage comes from the resuming of the previous stage (256-320-384), training details and iterations can be found at Sec.3.4 & Sec. 11.3 in https://arxiv.org/pdf/2502.03649.

2. The provided Docker image is exactly what we use for training and inferring results in the paper. The same results with paper should be obtained if inferring with the given Docker image and checkpoint.

3.  Training Instability. For the "non-finite norm" errors: this project is built on the [MMagic](https://github.com/open-mmlab/mmagic)  framework, which is generally very stable. However, since the entropy model part is built on [DCVC-FM](https://github.com/microsoft/DCVC/tree/main/DCVC-FM), there are a few things to keep in mind: 
* Due to patent constraints, we cannot release the self-implemented training code of DCVC-FM entropy model, the current code is released for inference only.
* If you are implementing training pipeline, the current round-clip operation in entropy model is likely the reason for the instability.   To fix this, consider making the round operation differentiable during training.
* Alternatively, you might consider replacing the spatial entropy model with a different open-source codec, while only keeping our encoder and decoder.

**Comment by YuKDseele** [comment]
> Hi,
> 
> Thanks for your interest in our work!
> 
> 1. all_in_one_S.py/all_in_one_L.py files are configurations of the final stage.  We adopt a progressive training strategy (Sec. 3.4), so later stage comes from the resuming of the previous stage (256-320-384), training details and iterations can be found at Sec.3.4 & Sec. 11.3 in https://arxiv.org/pdf/2502.03649.
> 2. The provided Docker image is exactly what we use for training and inferring results in the paper. The same results with paper should be obtained if inferring with the given Docker image and checkpoint.
> 3. Training Instability. For the "non-finite norm" errors: this project is built on the [MMagic](https://github.com/open-mmlab/mmagic)  framework, which is generally very stable. However, since the entropy model part is built on [DCVC-FM](https://github.com/microsoft/DCVC/tree/main/DCVC-FM), there are a few things to keep in mind:
> 
> * Due to patent constraints, we cannot release the self-implemented training code of DCVC-FM entropy model, the current code is released for inference only.
> * If you are implementing training pipeline, the current round-clip operation in entropy model is likely the reason for the instability.   To fix this, consider making the round operation differentiable during training.
> * Alternatively, you might consider replacing the spatial entropy model with a different open-source codec, while only keeping our encoder and decoder.

Hi authors, thank you again for your previous guidance. It was very helpful.  
I have resolved the earlier issue and can now run training in a preliminary way.  
I still have a few questions and would really appreciate your clarification:
﻿
1. Does the all-in-one method’s restoration performance rely on the **multi-rate training script/schedule** and the corresponding **entropy-model design**?  
If we switch to **single-rate training**, will image restoration quality (e.g., PSNR/MS-SSIM) be affected significantly?
﻿
2. Is the training objective simply the standard **rate-distortion loss**  
\[
\mathcal{L} = \lambda \cdot D(\hat{x}, x_{gt}) + R_y + R_z
\]
with **\(x_{gt}\)** as the target, without additional restoration-specific losses?
﻿
3. Is the data augmentation pipeline exactly as follows: random crop, horizontal/vertical flip, and random clean-input replacement with probability 0.2 (i.e., using \(x_{gt}\) as input for some samples)?
﻿
4. Is this RD objective generally applicable to other encoder/decoder architectures, or is it tightly coupled with your current design?  
In other words, if we replace the backbone while keeping this loss, should the same training principle still hold?
﻿
Thanks again for your time and help.

**Comment by ZeldaM1** [comment]
Hi,

Glad those suggestions can help! Please find details below.

1. **Single-rate training**. Theoretically, the restoration performance should be orthogonal to the multi-rate training design. The purpose of using the multi-rate training is to enable a single model to handle flexible bitrates.  But we maintain caution as we didn't conduct such an ablation study.

2. **Loss function**. 

Just the RD loss in ([Sec.3.4, Eq.4](https://arxiv.org/pdf/2502.03649)). No additional restoration-specific losses.

The adopted RD loss is a widely adopted loss function among image compression (eg, [mix](https://arxiv.org/abs/2303.14978),[devil](https://arxiv.org/abs/2203.08450)), not specifically designed for our backbone. The core training principle should still hold when replacing the backbone.
 
3. **Augmentation pipeline**. \
**Weather Degradation setting**: We use the real-world degraded pairs provided by the dataset. The augmentation is limited to the aforementioned geometric transformations. No synthetic degradation here.
**Gaussian Noise setting**:  Gaussian noise of different variances is applied on the clean input  $x_{gt}$ to simulate degraded input.

Hope the explanation above can address your concerns. Let me know if you have any more questions. :)
